### PR TITLE
ci: [NPM] scale test pipeline using KWOK

### PR DESCRIPTION
**Pull request by huntergregory**
**Reason for Change**:
Automate NPM scale tests.

**Issue Fixed**:

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
- [ ] includes documentation
- [ ] adds unit tests

**Notes**:
#### 1. Scale Up Large (~1K fake Pods, 30 real Pods)
- Create Fake + Real Pods.
- Label them.
- Add NetPols.

#### 2. Test NPM Bootup Latency and Connectivity
Must achieve initial connectivity within 15 minutes, and apply a new NetworkPolicy within 30 seconds.

- Restart NPM to see performance improvements from #1900.
- Check connectivity. Fail if a timeout is exceeded.
  - Create two pinger Pods.
  - Check initial connectivity. If unsuccessful, retry until timeout exceeded.
    - Allowed: pinger <-> pinger
    - Allowed: real <-> real
    - Blocked: real <-> pinger
  - Add NetPol and assert connectivity. If unsuccessful, retry until timeout exceeded.
    - Allowed: real <-> pinger

#### 3. General CRUD at Medium Scale  (~30 fake Pods, 10 real Pods)
- Delete previous namespaces (including Pods and Netpols).
- Recreate everything in step 1.
- Delete and re-add labels.
- Delete and re-create Pods.
- Delete and re-create NetPols.

#### 4. Test Connectivity After CRUD
Must achieve initial connectivity within 10 minutes, and apply a new NetworkPolicy within 20 seconds.

- Check connectivity like in step 2.

### CODE REVIEW

File: .pipelines/npm/npm-conformance-tests.yaml [hunk 1]
```diff
@@ -90,14 +90,14 @@ jobs:
...
       matrix:
-        v1-default:
-          AZURE_CLUSTER: "conformance-v1-default"
-          PROFILE: "v1-default"
-          IS_STRESS_TEST: "false"
-        v2-default:
-          AZURE_CLUSTER: "conformance-v2-default"
-          PROFILE: "v2-default"
-          IS_STRESS_TEST: "false"
+        # v1-default:
```
Comment (huntergregory) [inline .pipelines/npm/npm-conformance-tests.yaml:93]: temporarily comment out to limit resources used while testing

File: .pipelines/npm/npm-conformance-tests.yaml [hunk 2]
```diff
@@ -345,14 +345,14 @@ jobs:
           done
 
           exit $exitCode
-        displayName: "Run Test Suite and Get Logs"
+        displayName: "Run Test Suite and Get Logs - $(PROFILE)"
```
Comment (huntergregory) [inline .pipelines/npm/npm-conformance-tests.yaml:348]: this change lets us distinguish job failures in our pipeline report

File: .pipelines/npm/npm-conformance-tests.yaml [hunk 3]
```diff
@@ -463,7 +463,169 @@ jobs:
         name: cyclonus
         displayName: "Run Cyclonus Test"
         failOnStderr: false
+        condition: succeeded()
+
+      - bash: |
```
Comment (huntergregory) [inline .pipelines/npm/npm-conformance-tests.yaml:468]: diff says we're adding this block, but really we're just copying it from here to below the scale test

**Review by vakalapa — DISMISSED** [review]
Make sure you add times on why it is taking almost 2hrs for the connectivity tests to start ?

**Comment by huntergregory** [comment]
TODO: make new pipeline instead of using conformance

**Comment by huntergregory** [comment]
> Make sure you add times on why it is taking almost 2hrs for the connectivity tests to start ?

Fixed. Kwok wasn't running, so the fake Pods weren't being scheduled

File: .pipelines/npm/npm-scale-test.yaml [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+            kubectl apply -f azure-npm.yaml
+
+            # swap azure-npm image with one built during run
+            kubectl set image daemonset/azure-npm-win -n kube-system azure-npm=$IMAGE_REGISTRY/azure-npm:windows-amd64-ltsc2022-$(TAG)
+
+            sleep 5s
+            echo "waiting for NPM to start running..."
+            kubectl wait --for=condition=Ready pod -l k8s-app=azure-npm -n kube-system --timeout=20m
+            echo "sleep 1m to let NPM restart in case of bootup failure due to HNS errors"
+            sleep 1m
```
Comment (vakalapa) [inline .pipelines/npm/npm-scale-test.yaml:183]: windows pods can take longer than 1m sometimes it think ?

File: .pipelines/npm/npm-scale-test.yaml [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+        inputs:
+          azureSubscription: $(BUILD_VALIDATIONS_SERVICE_CONNECTION)
+          scriptType: "bash"
+          scriptLocation: "inlineScript"
+          failOnStderr: true
+          condition: succeeded()
+          inlineScript: |
+            set -e
+            mkdir -p $(System.DefaultWorkingDirectory)/$(RESOURCE_GROUP)-$(PROFILE)
+            ./kwok --kubeconfig ~/.kube/config \
```
Comment (matmerr) [inline .pipelines/npm/npm-scale-test.yaml:206]: this is pretty useful for other scenarios, maybe after this add a make recipe to attach kwok nodes

**Comment by huntergregory** [comment]
going to hold off on merging until HNS fixes are in prod (probably next week)

**Review by vakalapa — APPROVED** [review]

**Pull request merged**